cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,472 @@
+        int leaderReplicaCount = broker.numLeadersFor(mustHaveLeaderReplicaPerBrokerTopicName);
+        if (leaderReplicaCount < minTopicLeadersPerBroker()) {
+          throw new OptimizationFailureException(String.format("Broker %d does not have enough leader replica for topic %s. "
+              + "Minimum required per-broker leader replica count %d. Actual broker leader replica count %d",
+              broker.id(), mustHaveLeaderReplicaPerBrokerTopicName, minTopicLeadersPerBroker(), leaderReplicaCount));
+        }
+      }
+    }
+  }
+
+  /**
+   * Rebalance the given broker without violating the constraints of the current goal and optimized goals.
+   * @param broker         Broker to be balanced.
+   * @param clusterModel   The state of the cluster.
+   * @param optimizedGoals Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker,
+                                    ClusterModel clusterModel,
+                                    Set<Goal> optimizedGoals,
+                                    OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
+    moveAwayOfflineReplicas(broker, clusterModel, optimizedGoals, optimizationOptions);
+    if (_mustHaveTopicLeadersPerBroker.isEmpty()) {
+      return; // Early termination to avoid some unnecessary computation
+    }
+    if (!isEligibleToHaveLeaders(broker, optimizationOptions)) {
+      return;
+    }
+    for (String topicMustHaveLeaderReplicaPerBroker : _mustHaveTopicLeadersPerBroker) {
+      maybeMoveLeaderReplicaOfTopicToBroker(topicMustHaveLeaderReplicaPerBroker, broker, clusterModel, optimizedGoals, optimizationOptions);
+    }
+  }
+
+  private void maybeMoveLeaderReplicaOfTopicToBroker(String topicMustHaveLeaderReplicaPerBroker,
+                                                     Broker broker,
+                                                     ClusterModel clusterModel,
+                                                     Set<Goal> optimizedGoals,
+                                                     OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    if (brokerHasSufficientLeaderReplicasOfTopic(broker, topicMustHaveLeaderReplicaPerBroker)) {
+      return; // This broker has enough leader replica(s)
+    }
+    // Try to elect follower replica(s) of the interested topic on this broker to be leader
+    Set<Replica> followerReplicas = broker.trackedSortedReplicas(_replicaSortName)
+                                          .sortedReplicas(false).stream().filter(
+                                              replica -> !replica.isLeader()
+                                                         && replica.topicPartition().topic().equals(topicMustHaveLeaderReplicaPerBroker))
```

Comment:
Don't we already `selectReplicasBasedOnIncludedTopics`? Why do we need this filtering?

* There is a similar use below. Please also verify that.

Reply:
The `selectReplicasBasedOnIncludedTopics` method takes a set of topic names. Here, we need to find (leader) replicas of a specific topic with the name `topicMustHaveLeaderReplicaPerBroker`. So, this filtering is necessary.